tests/test_collection.py

```diff
@@ -0,0 +1,366 @@
+
+    def test_optional_sequence(self):
+        @dataclass
+        class IntOptionalSequence:
+            value: Optional[Sequence[int]]
+
+        schema = IntOptionalSequence.Schema()
+        for data_in in ({"value": list(range(5))}, {"value": []}, {"value": None}):
+            self.assertEqual(schema.dump(schema.load(data_in)), data_in)
+
+    def test_sequence_of_optional(self):
+        @dataclass
+        class OptionalIntSequence:
+            value: Sequence[Optional[int]]
+
+        schema = OptionalIntSequence.Schema()
+        for data_in in ({"value": list(range(5))}, {"value": [1, None, 2, None]}):
+            self.assertEqual(schema.dump(schema.load(data_in)), data_in)
+
+    def test_nested_sequence(self):
+        @dataclass
+        class SequenceIntSequence:
+            value: Sequence[Sequence[int]]
+
+        schema = SequenceIntSequence.Schema()
+        for data_in in (
+            {"value": []},
+            {"value": [[], []]},
+            {"value": [[1, 2, 3], [4, 5, 6]]},
+        ):
+            self.assertEqual(schema.dump(schema.load(data_in)), data_in)
+
+    def test_sequence_of_dataclass(self):
+        @dataclass
+        class Elm:
+            value: int
+
+        @dataclass
+        class SequenceIntSequence:
+            value: Sequence[Elm]
+
+        schema = SequenceIntSequence.Schema()
+        for data_in in (
+            {"value": []},
+            {"value": [{"value": 1}]},
+        ):
+            self.assertEqual(schema.dump(schema.load(data_in)), data_in)
+
+    def test_sequence_of_optional_dataclass(self):
+        @dataclass
+        class Elm:
+            value: int
+
+        @dataclass
+        class SequenceIntSequence:
+            value: Sequence[Optional[Elm]]
+
+        schema = SequenceIntSequence.Schema()
+        for data_in in (
+            {"value": []},
+            {"value": [None]},
+            {"value": [{"value": 1}]},
+        ):
+            self.assertEqual(schema.dump(schema.load(data_in)), data_in)
+
+    def test_serializing_sets_as_sequence_is_possible_but_risky(self):
+        @dataclass
+        class StrSequence:
+            value: Sequence[str]
+
+        str_schema = StrSequence.Schema()
+
+        # you can dump it, but order will be random
+        sequence = (
+            "this",
+            "has",
+            "only",
+            "1",
+            "chance",
+            "in",
+            "40M",
+            "to",
+            "keep",
+            "it's",
+            "ordered",
+        )
+        dump = str_schema.dump(StrSequence(value=set(sequence)))
+        self.assertNotEqual(
+            dump,
+            {"value": sequence},
+        )
+
+        @dataclass
+        class IntSequence:
+            value: Sequence[int]
+
+        int_schema = IntSequence.Schema()
+
+        # and to make it worse, int sets are a valid sequence that keep order....
+        dump = int_schema.dump(IntSequence(value={1, 2, 3, 4, 5}))
```

Comment:
Is that guaranteed by the spec or just an implementation detail ?

Reply:
It's more a find-out while testing. IMO it comes down to how a set is implemented were ints will stay in order, i suppose because d`hash(int)=int`.

But again, looking back, this whole test seams out of scope. I will remove it